File: AVS_KT3/shrub.py
import random
import string


# Куст
# ----------------------------------------------

# Месяц цветения
monthOfFlowering = {1: "January", 2: "February", 3: "March", 4: "April", 5: "May", 6: "June", 7: "July",
                    8: "August", 9: "September", 10: "October", 11: "November", 12: "December"}


# Считывание данных из файла
def Read(shrub, strArray, i):
    # Должно быть как минимум два непрочитанных значения в массиве (название и номер месяца цветения)
    if i >= len(strArray) - 1:
        return 0
    shrub.append("shrub")
    shrub.append(strArray[i])
    shrub.append(monthOfFlowering[int(strArray[i + 1])])
    i += 2
    return i


# Генерация рандомного названия
def RandomName(length):
    letters = string.ascii_letters
    rand_string = ''.join(random.choice(letters) for i in range(length))
    return rand_string


# Заполнение списка случайными параметрами
def Random(shrub):
    shrub.append("shrub")
    shrub.append(RandomName(random.randrange(1, 20, 1)))
    shrub.append(monthOfFlowering[random.randrange(1, 12, 1)])
    pass


# Вывод информации на экран
def Print(shrub):
    print("Shrub name: ", shrub[1], ", month of flowering: ", shrub[2], ", percent of vowel letters = ",
          round(PercentageOfVowelLetters(shrub) * 100, 3), "%", sep="")
    pass


# Вывод информации в поток
def Write(shrub, ostream):
    ostream.write("Shrub name: {}, month of flowering: {}, percent of vowel letters = {}%".format(shrub[1],
                  shrub[2], round(PercentageOfVowelLetters(shrub) * 100, 3)))
    pass


# Подсчет процента гласных букв по отношению ко всему количеству
def PercentageOfVowelLetters(shrub):
    countVowelLetters = 0
    for letter in shrub[1]:
        if letter == 'A' or letter == 'E' \
                or letter == 'I' or letter == 'O' \
                or letter == 'U' or letter == 'Y' \
                or letter == 'a' or letter == 'e' \
                or letter == 'i' or letter == 'o' \
                or letter == 'u' or letter == 'y':
            countVowelLetters += 1
    return float(countVowelLetters) / len(shrub[1])
    pass

File: AVS_KT1/Source/shrub.cpp
//------------------------------------------------------------------------------
// shrub.cpp - содержит процедуру ввода параметров
// для созданного кустарника
//------------------------------------------------------------------------------

#include "../Headers/shrub.h"

//------------------------------------------------------------------------------
// Ввод параметров кустарника из файла
void In(shrub& shrub, ifstream& ifstream) {
    ifstream >> shrub.name;

    int typeOfFlower;
    ifstream >> typeOfFlower;

    switch (typeOfFlower) {
        case 1:
            shrub.month = shrub.January;
            return;
        case 2:
            shrub.month = shrub.February;
            return;
        case 3:
            shrub.month = shrub.March;
            return;
        case 4:
            shrub.month = shrub.April;
            return;
        case 5:
            shrub.month = shrub.May;
            return;
        case 6:
            shrub.month = shrub.June;
            return;
        case 7:
            shrub.month = shrub.July;
            return;
        case 8:
            shrub.month = shrub.August;
            return;
        case 9:
            shrub.month = shrub.September;
            return;
        case 10:
            shrub.month = shrub.October;
            return;
        case 11:
            shrub.month = shrub.November;
            return;
        case 12:
            shrub.month = shrub.December;
            return;
        default:
            return;
    }
}

// Случайный ввод параметров кустарника
void InRnd(shrub& shrub) {
    int countLetters = Random() - 1;
    for (unsigned int i = 0; i < countLetters; ++i){
        shrub.name[i] = (char)(Random() + 65);
    }

    shrub.name[countLetters] = '\0';

    int monthOfFlowering = Random() % 12;

    switch (monthOfFlowering) {
        case 0:
            shrub.month = shrub.January;
            return;
        case 1:
            shrub.month = shrub.February;
            return;
        case 2:
            shrub.month = shrub.March;
            return;
        case 3:
            shrub.month = shrub.April;
            return;
        case 4:
            shrub.month = shrub.May;
            return;
        case 5:
            shrub.month = shrub.June;
            return;
        case 6:
            shrub.month = shrub.July;
            return;
        case 7:
            shrub.month = shrub.August;
            return;
        case 8:
            shrub.month = shrub.September;
            return;
        case 9:
            shrub.month = shrub.October;
            return;
        case 10:
            shrub.month = shrub.November;
            return;
        case 11:
            shrub.month = shrub.December;
            return;
        default:
            return;
    }
}

//------------------------------------------------------------------------------
// Вывод параметров кустарнкиа в форматируемый поток
void Out(shrub& shrub, ofstream& ofstream) {
    string monthOfFlowering;

    if (shrub.month == shrub.January) {
        monthOfFlowering = "январь";
    }
    else if (shrub.month == shrub.February) {
        monthOfFlowering = "февраль";
    }
    else if (shrub.month == shrub.March) {
        monthOfFlowering = "март";
    }
    else if (shrub.month == shrub.April) {
        monthOfFlowering = "апрель";
    }
    else if (shrub.month == shrub.May) {
        monthOfFlowering = "май";
    }
    else if (shrub.month == shrub.June) {
        monthOfFlowering = "июнь";
    }
    else if (shrub.month == shrub.July) {
        monthOfFlowering = "июль";
    }
    else if (shrub.month == shrub.August) {
        monthOfFlowering = "август";
    }
    else if (shrub.month == shrub.September) {
        monthOfFlowering = "сентябрь";
    }
    else if (shrub.month == shrub.October) {
        monthOfFlowering = "октябрь";
    }
    else if (shrub.month == shrub.November) {
        monthOfFlowering = "ноябрь";
    }
    else if (shrub.month == shrub.December) {
        monthOfFlowering = "декабрь";
    }

    ofstream << "Название кустарника: " << shrub.name
             << "\nМесяц цветения: " << monthOfFlowering
             << "\nПроцент гласных букв по отношению ко всем буквам в названии = " << PercentageOfVowelLetters(shrub)
             << "%\n";
}

//------------------------------------------------------------------------------
// Вычисление процента гласных букв по отношению ко всем буквам в названии
double PercentageOfVowelLetters(shrub& shrub) {
    double countVowelLetters = 0;
    double countLetters = 0;

    for (char i : shrub.name) {
        if (i != ' ') {
            countLetters++;
            if (i == 'A' || i == 'E'
            || i == 'I' || i == 'O'
            || i == 'U' || i == 'Y'
            || i == 'a' || i == 'e'
            || i == 'i' || i == 'o'
            || i == 'u' || i == 'y') {
                countVowelLetters++;
            }
        }
    }

    return (countVowelLetters / countLetters) * 100;
}
